File: JAVA/socketry-main/src/main/java/com/socketry/packetparser/Packet.java
package com.socketry.packetparser;

import java.nio.ByteBuffer;
import java.nio.channels.WritableByteChannel;
import java.util.Arrays;


record CREPacket(byte fnId, byte callId, byte[] arguments) {
    public static CREPacket parse(byte[] data) {
        byte fnId = data[1];
        byte callId = data[2];
        int argumentsLength = data.length - 3;
        byte[] arguments = new byte[argumentsLength];
        System.arraycopy(data, 3, arguments, 0, argumentsLength);
        return new CREPacket(fnId, callId, arguments);
    }
}

public sealed interface Packet permits Packet.Call, Packet.Result, Packet.Error, Packet.Init, Packet.Accept, Packet.Ping, Packet.Pong {

    static Packet parse(byte[] data) {
        byte type = data[0];
        switch (type) {
            case PacketType.CALL:
                return Call.parse(data);
            case PacketType.RESULT:
                return Result.parse(data);
            case PacketType.ERROR:
                return Error.parse(data);
            case PacketType.INIT:
                return new Init(Arrays.copyOfRange(data, 1, data.length));
            case PacketType.ACCEPT:
                return Accept.parse(data);
            case PacketType.PING:
                return Ping.INSTANCE;
            case PacketType.PONG:
                return Pong.INSTANCE;
            default:
                throw new IllegalArgumentException("Unknown packet type: " + type);
        }
    }
    
    static ByteBuffer serialize(Packet packet) {
        ByteBuffer buffer = ByteBuffer.allocate(1 + switch (packet) {
            case Packet.Call call -> 2 + call.arguments().length;
            case Packet.Result result -> 2 + result.response().length;
            case Packet.Error error -> 2 + error.error().length;
            case Packet.Init init -> init.channels().length;
            case Packet.Accept accept -> 2 * accept.ports().length;
            case Packet.Ping ping -> 0;
            case Packet.Pong pong -> 0;
        });
        switch (packet) {
            case Packet.Call call:
                buffer.put(PacketType.CALL);
                buffer.put(call.fnId());
                buffer.put(call.callId());
                buffer.put(call.arguments());
                break;
            case Packet.Result result:
                buffer.put(PacketType.RESULT);
                buffer.put(result.fnId());
                buffer.put(result.callId());
                buffer.put(result.response());
                break;
            case Packet.Error error:
                buffer.put(PacketType.ERROR);
                buffer.put(error.fnId());
                buffer.put(error.callId());
                buffer.put(error.error());
                break;
            case Packet.Init init:
                buffer.put(PacketType.INIT);
                buffer.put(init.channels());
                break;
            case Packet.Accept accept:
                buffer.put(PacketType.ACCEPT);
                for (short port : accept.ports()) {
                    buffer.putShort(port); // BIG ENDIAN
                }
                break;
            case Packet.Ping ping:
                buffer.put(PacketType.PING);
                break;
            case Packet.Pong pong:
                buffer.put(PacketType.PONG);
                break;
        }
        return buffer;
    }

    ThreadLocal<ByteBuffer> BUFFER_CACHE = 
        ThreadLocal.withInitial(() -> ByteBuffer.allocate(1024*1024*1024)); // 1MB

    static ByteBuffer serializeFast(Packet packet) {
        int size = switch (packet) {
            case Packet.Call call -> 2 + call.arguments().length;
            case Packet.Result result -> 2 + result.response().length;
            case Packet.Error error -> 2 + error.error().length;
            case Packet.Init init -> init.channels().length;
            case Packet.Accept accept -> 2 * accept.ports().length;
            case Packet.Ping ping -> 0;
            case Packet.Pong pong -> 0;
        };
        ByteBuffer buffer = BUFFER_CACHE.get();
        buffer.clear();
        buffer.putInt(size + 1);
        switch (packet) {
            case Packet.Call call:
                buffer.put(PacketType.CALL);
                buffer.put(call.fnId());
                buffer.put(call.callId());
                buffer.put(call.arguments());
                break;
            case Packet.Result result:
                buffer.put(PacketType.RESULT);
                buffer.put(result.fnId());
                buffer.put(result.callId());
                buffer.put(result.response());
                break;
            case Packet.Error error:
                buffer.put(PacketType.ERROR);
                buffer.put(error.fnId());
                buffer.put(error.callId());
                buffer.put(error.error());
                break;
            case Packet.Init init:
                buffer.put(PacketType.INIT);
                buffer.put(init.channels());
                break;
            case Packet.Accept accept:
                buffer.put(PacketType.ACCEPT);
                for (short port : accept.ports()) {
                    buffer.putShort(port); // BIG ENDIAN
                }
                break;
            case Packet.Ping ping:
                buffer.put(PacketType.PING);
                break;
            case Packet.Pong pong:
                buffer.put(PacketType.PONG);
                break;
        }
        buffer.flip();
        return buffer;
    }


    record Call(byte fnId, byte callId, byte[] arguments) implements Packet {
        static Call parse(byte[] data) {
            CREPacket crePacket = CREPacket.parse(data);
            return new Call(crePacket.fnId(), crePacket.callId(), crePacket.arguments());
        }
    }

    record Result(byte fnId, byte callId, byte[] response) implements Packet {
        static Result parse(byte[] data) {
            CREPacket crePacket = CREPacket.parse(data);
            return new Result(crePacket.fnId(), crePacket.callId(), crePacket.arguments());
        }
    }

    record Error(byte fnId, byte callId, byte[] error) implements Packet {
        static Error parse(byte[] data) {
            CREPacket crePacket = CREPacket.parse(data);
            return new Error(crePacket.fnId(), crePacket.callId(), crePacket.arguments());
        }
    }

    record Init(byte[] channels) implements Packet {}

    record Accept(short[] ports) implements Packet {
        static Accept parse(byte[] data) {
            // maybe a faster 0copy method exists, but i digress.
            short[] ports = new short[(data.length - 1) / 2]; // assume length is odd (even + 1 for type byte)
            for (int i = 1; i < data.length; i += 2) {
                ports[i / 2] = (short) (((data[i] & 0xFF) << 8) | (data[i + 1] & 0xFF)); // BIG ENDIAN
            }
            return new Accept(ports);
        }
    }

    final class Ping implements Packet {
        public static final Ping INSTANCE = new Ping();
        private Ping() {}
    }

    final class Pong implements Packet {
        public static final Pong INSTANCE = new Pong();
        private Pong() {}
    }
}

File: C#/Socketry/packetparser/Packet.cs
namespace packetparser
{
    /// <summary>
    /// Function to create a RPC packet.
    /// </summary>
    /// <param name="fnId">the function id</param>
    /// <param name="callId">the call id</param>
    /// <param name="arguments">the arguments to pass</param>
    record CREPacket(byte fnId, byte callId, byte[] arguments)
    {
        public static CREPacket Parse(byte[] data)
        {
            byte fnId = data[1];
            byte callId = data[2];
            int argumentsLength = data.Length - 3;
            byte[] arguments = new byte[argumentsLength];
            Array.Copy(data, 3, arguments, 0, argumentsLength);
            return new CREPacket(fnId, callId, arguments);
        }
    }
    /// <summary>
    /// Interface for Packet.
    /// </summary>
    public interface Packet
    {
        /// <summary>
        /// The function to parse a given packet.
        /// </summary>
        /// <param name="data">the data to be parsed</param>
        /// <returns></returns>
        /// <exception cref="ArgumentException"></exception>
        static Packet Parse(byte[] data)
        {
            byte type = data[0];
            switch (type)
            {
                case PacketType.CALL:
                    return Call.Parse(data);
                case PacketType.RESULT:
                    return Result.Parse(data);
                case PacketType.ERROR:
                    return Error.Parse(data);
                case PacketType.INIT:
                    int length = data.Length - 1;
                    byte[] packetData = new byte[length];
                    Array.Copy(data, 1, packetData, 0, length);
                    return new Init(packetData);
                case PacketType.ACCEPT:
                    return Accept.Parse(data);
                case PacketType.PING:
                    return Ping.Instance;
                case PacketType.PONG:
                    return Pong.Instance;
                default:
                    throw new ArgumentException($"Unknown packet type {type}");
            }
        }

        /// <summary>
        /// Function to create new packets.
        /// </summary>
        /// <param name="packet">the type of packet to create</param>
        /// <returns>The new packet with values</returns>
        /// <exception cref="ArgumentOutOfRangeException"></exception>
        static MemoryStream Serialize(Packet packet)
        {
            int size = 1 + packet switch
            {
                Packet.Call call => 2 + call.arguments.Length,
                Packet.Result result => 2 + result.response.Length,
                Packet.Error error => 2 + error.error.Length,
                Packet.Init init => init.channels.Length,
                Packet.Accept accept => 2 * accept.ports.Length,
                Packet.Ping => 0,
                Packet.Pong => 0,
                _ => throw new ArgumentOutOfRangeException()
            };
            MemoryStream buffer = new MemoryStream(size);
            BinaryWriter writer = new BinaryWriter(buffer);

            switch (packet)
            {
                case Packet.Call call:
                    writer.Write(PacketType.CALL);
                    writer.Write(call.fnId);
                    writer.Write(call.callId);
                    writer.Write(call.arguments);
                    break;
                case Packet.Result result:
                    writer.Write(PacketType.RESULT);
                    writer.Write(result.fnId);
                    writer.Write(result.callId);
                    writer.Write(result.response);
                    break;
                case Packet.Error error:
                    writer.Write(PacketType.ERROR);
                    writer.Write(error.fnId);
                    writer.Write(error.callId);
                    writer.Write(error.error);
                    break;
                case Packet.Init init:
                    writer.Write(PacketType.INIT);
                    writer.Write(init.channels);
                    break;
                case Packet.Accept accept:
                    writer.Write(PacketType.ACCEPT);
                    foreach (short port in accept.ports)
                    {
                        writer.Write(port);
                    }
                    break;
                case Packet.Ping ping:
                    writer.Write(PacketType.PING);
                    break;
                case Packet.Pong pong:
                    writer.Write(PacketType.PONG);
                    break;

            }
            return buffer;
        }

        static ThreadLocal<MemoryStream> BUFFER_CACHE =
    new ThreadLocal<MemoryStream>(() => new MemoryStream(1024 * 1024 * 1024));

        /// <summary>
        /// Function to creaet packet form the cache.
        /// </summary>
        /// <param name="packet">The packet to be created</param>
        /// <returns>The new packet</returns>
        /// <exception cref="ArgumentOutOfRangeException"></exception>
        static MemoryStream serializeFast(Packet packet)
        {
            int size = 1 + packet switch
            {
                Packet.Call call => 2 + call.arguments.Length,
                Packet.Result result => 2 + result.response.Length,
                Packet.Error error => 2 + error.error.Length,
                Packet.Init init => init.channels.Length,
                Packet.Accept accept => 2 * accept.ports.Length,
                Packet.Ping => 0,
                Packet.Pong => 0,
                _ => throw new ArgumentOutOfRangeException()
            };
            MemoryStream buffer = BUFFER_CACHE.Value!;
            BinaryWriter writer = new BinaryWriter(buffer);
            buffer.Position = 0;
            switch (packet)
            {
                case Packet.Call call:
                    writer.Write(PacketType.CALL);
                    writer.Write(call.fnId);
                    writer.Write(call.callId);
                    writer.Write(call.arguments);
                    break;
                case Packet.Result result:
                    writer.Write(PacketType.RESULT);
                    writer.Write(result.fnId);
                    writer.Write(result.callId);
                    writer.Write(result.response);
                    break;
                case Packet.Error error:
                    writer.Write(PacketType.ERROR);
                    writer.Write(error.fnId);
                    writer.Write(error.callId);
                    writer.Write(error.error);
                    break;
                case Packet.Init init:
                    writer.Write(PacketType.INIT);
                    writer.Write(init.channels);
                    break;
                case Packet.Accept accept:
                    writer.Write(PacketType.ACCEPT);
                    foreach (short port in accept.ports)
                    {
                        writer.Write(port);
                    }
                    break;
                case Packet.Ping ping:
                    writer.Write(PacketType.PING);
                    break;
                case Packet.Pong pong:
                    writer.Write(PacketType.PONG);
                    break;
            }
            buffer.SetLength(buffer.Position);
            buffer.Position = 0;
            return buffer;
        }

        /// <summary>
        /// Function to create a CALL packet
        /// </summary>
        /// <param name="fnId"> the function id</param>
        /// <param name="callId"> the call id </param>
        /// <param name="arguments"> the arguments to pass </param>
        record Call(byte fnId, byte callId, byte[] arguments) : Packet
        {
            public static Call Parse(byte[] data)
            {
                CREPacket crepacket = CREPacket.Parse(data);
                return new Call(crepacket.fnId, crepacket.callId, crepacket.arguments);
            }
        }

        /// <summary>
        /// Function to create a RESULT packet
        /// </summary>
        /// <param name="fnId"> the function id</param>
        /// <param name="callId"> the call id </param>
        /// <param name="response"> the response of the arguments </param>
        record Result(byte fnId, byte callId, byte[] response) : Packet
        {
            public static Result Parse(byte[] data)
            {
                CREPacket crepacket = CREPacket.Parse(data);
                return new Result(crepacket.fnId, crepacket.callId, crepacket.arguments);
            }
        }

        /// <summary>
        /// Function to create a ERROR packet
        /// </summary>
        /// <param name="fnId"> the function id</param>
        /// <param name="callId"> the call id </param>
        /// <param name="error"> the error of the packet </param>
        record Error(byte fnId, byte callId, byte[] error) : Packet
        {
            public static Error Parse(byte[] data)
            {
                CREPacket crepacket = CREPacket.Parse(data);
                return new Error(crepacket.fnId, crepacket.callId, crepacket.arguments);
            }
        }

        /// <summary>
        /// Function to create a INIT packet
        /// </summary>
        /// <param name="channels"> the number of channels </param>
        record Init(byte[] channels) : Packet { }

        /// <summary>
        /// Function to create a ACCEPT packet
        /// </summary>
        /// <param name="ports"> the list of ports </param>
        record Accept(short[] ports) : Packet
        {
            public static Accept Parse(byte[] data)
            {
                short[] ports = new short[(data.Length - 1) / 2]; // assume length is odd (even + 1 for type byte)
                for (int i = 1; i < data.Length; i += 2)
                {
                    ports[i / 2] = (short)(((data[i] & 0xFF) << 8) | (data[i + 1] & 0xFF)); // BIG ENDIAN
                }
                return new Accept(ports);
            }
        }

        /// <summary>
        /// Function to create a PING packet
        /// </summary>
        class Ping : Packet
        {
            public static Ping Instance = new Ping();
            private Ping() { }
        }

        /// <summary>
        /// Function to create a PONG packet
        /// </summary>
        class Pong : Packet
        {
            public static Pong Instance = new Pong();
            private Pong() { }
        }
    }
}
